frame/election-provider-multi-phase/src/migrations.rs

```diff
@@ -0,0 +1,69 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod v1 {
+	use frame_support::{
+		storage::unhashed,
+		traits::{Defensive, GetStorageVersion, OnRuntimeUpgrade},
+		BoundedVec,
+	};
+	use sp_std::collections::btree_map::BTreeMap;
+
+	use crate::*;
+	pub struct MigrateToV1<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV1<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let current = Pallet::<T>::current_storage_version();
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			log!(
+				info,
+				"Running migration with current storage version {:?} / onchain {:?}",
+				current,
+				onchain
+			);
+
+			if current == 1 && onchain == 0 {
```

Comment:
Probably the condition could just be `if onchain == 0`. Pallet version is in a way implicit since its v1 migrations.

Reply:
can't agree with this, what happens if we run v2 directly on some chain?

Reply:
Yes, v0 -> v2 or more will most likely involve additional / different migrations (after all, why else would we create another version).

Reply:
Reading again, it makes sense since we store the (latest) pallet version in the storage. But with this pattern, we will never be able to do multiple migration in one upgrade, is that correct?

My instinctive expectation was, given onchain is 2 versions behind, v1 runs, checks precondition `onchain == 0`, sets onchain version to 1, v2 runs, checks precondition `onchain == 1`, sets onchain version to 2.  This is similar to what we have in staking migration.

Is there a reason why we are following different migration patterns in different places? Since this is the first migration for this pallet, is this the preferred pattern of choice today?

Reply:
I normally only do `StorageVersion::get::<Pallet<T>>() == old_version`. Why do you even need these two checks?

Reply:
Yeah, when/if we need to run two migrations - it's going to be a PITA. This is a downside of current migration approach.
What would work instead

V1_MIGRATION_VERSION = 1
if onchain == 0 && V1_MIGRAION_VERSION <= current {
proceed
}

And so forth.

Reply:
What this achieves is we can limit the upper version to `current`.
If I'm being completely honest - I'm not sure why `CURRENT` exists at all. It would make sense
if we were running migrations automatically as opposed to specifying them in the executive.

Reply:
It needs to be done via executive since I believe the runtime can have explicit control over migration and not to run into unpredictable migrations. I am not even sure how it can work without executive.

Reply:
We could probably make these checks part of the `OnRuntimeMigration<OldVersion, NewVersion>` or whatever so we dont duplicate this code every time and wonder what it does :laughing:

Reply:
> We could probably make these checks part of the `OnRuntimeMigration<OldVersion, NewVersion>` or whatever so we dont duplicate this code every time and wonder what it does 😆

Yes, that would be amazing. I guess the only issue we have with this now is that some of the migrations are still using enums and stuff like that. Could be great to flesh it out in a ticket and make sure we do a refactoring to support this. Could be a big one though. 
Or otherwise just enable this functionality for all the pallets that have current/onchain.

Reply:
Interesting thread, let's continue in https://github.com/paritytech/substrate/issues/10308 or a fresh issue about automatic migrations.